tools/src/main/java/org/terracotta/utilities/io/Files.java

```diff
@@ -0,0 +1,1856 @@
+    Set<CopyOption> copyOptions = new LinkedHashSet<>(Arrays.asList(options));
+    copyOptions.add(NOFOLLOW_LINKS);
+    copyOptions.add(COPY_ATTRIBUTES);
+    copyOptions.add(ExtendedOption.RECURSIVE);
+
+    try {
+      return rename(source, target);      // If rename was successful, we're done
+    } catch (AtomicMoveNotSupportedException e) {
+      // Handled below
+    }
+
+    copy(source, target, copyOptions.toArray(new CopyOption[0]));
+    deleteTree(source);
+    return target;
+  }
+
+  /**
+   * Delete the file or directory tree using a rename/delete scheme with retry for
+   * {@code FileSystemException} instances indicating interference from temporary access by other processes.
+   *
+   * @param path the file or root of the directory to delete
+   * @param retryDirNotEmpty enable retry for {@link DirectoryNotEmptyException}
+   * @param renameTimeLimit the time limit to apply to renaming the file/directory before deletion
+   *
+   * @throws FileSystemException for a retryable exception if the time permitted for rename attempts
+   *          and retryable exception handling is exhausted or interrupted
+   * @throws IllegalArgumentException if {@code renameTimeLimit} is not at least twice the operation repeat delay
+   * @throws IOException if the rename or delete fails
+   * @throws SecurityException if permission to access the file/directory to rename/delete is not sufficient
+   */
+  private static void deleteTreeWithRetry(Path path, boolean retryDirNotEmpty, Duration renameTimeLimit)
+      throws IOException {
+    if (LOGGER.isDebugEnabled()) {
+      LOGGER.debug("[deleteTreeWithRetry] Deleting \"{}\"", path);
+    }
+
+    /*
+     * First, attempt a rename of the file/directory. The expectation with this move is that it's a
+     * "simple" rename -- we're not moving across file system roots -- so an "atomic" operation is
+     * what we want. If this fails for an AtomicMoveNotSupportedException, it gets reflected to the
+     * caller as a delete failure.
+     *
+     * If the rename succeeds, the "delete" is successful and the caller's following operations can
+     * proceed,  The remainder of the delete work can be handled in the background if warranted.
+     */
+    Path renamedPath = retryingRename(path, () -> randomName("del"), renameTimeLimit);
+
+    /*
+     * Deleting a file or a directory and its contents.  In this, links are not followed -- only
+     * direct tree content is removed.
+     */
+    LOGGER.debug("Deleting \"{}\" renamed from \"{}\"", renamedPath, path);
+    deleteTreeWithBackgroundRetry(renamedPath, retryDirNotEmpty);
+  }
+
+  /**
+   * Attempts to delete a file/directory retrying in background in the event of a retryable failure.
+   *
+   * @param path the {@code Path} to delete
+   * @param retryDirNotEmpty enable retry for {@link DirectoryNotEmptyException}
+   */
+  private static void deleteTreeWithBackgroundRetry(Path path, boolean retryDirNotEmpty) {
```

Comment:
One thing I could add to this scheme is to add each file/directory to the 'deleteOnExit' queue in the event we give up on trying to delete.  The code would go into `RetryingDeleteTask` and traverse the tree as `deleteTree` does.